File: sdks/dotnet/ArcMiddleware/src/ArcIdentityExtractor.cs
// Default identity extraction from HTTP request headers.
//
// Mirrors the Rust extract_caller logic in arc-api-protect.
// Extracts caller identity from Authorization headers, API keys, and cookies.

using System.Security.Cryptography;
using System.Text;
using Microsoft.AspNetCore.Http;

namespace Backbay.Arc;

/// <summary>
/// Delegate for extracting caller identity from an HTTP request.
/// </summary>
public delegate CallerIdentity IdentityExtractorDelegate(HttpRequest request);

/// <summary>
/// Default identity extraction utilities.
/// </summary>
public static class ArcIdentityExtractor
{
    /// <summary>
    /// Compute SHA-256 hex digest of a string.
    /// </summary>
    public static string Sha256Hex(string input)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(input));
        return Convert.ToHexStringLower(bytes);
    }

    /// <summary>
    /// Default identity extractor. Checks headers in order:
    /// 1. Authorization: Bearer token
    /// 2. X-API-Key header
    /// 3. Cookie header
    /// 4. Anonymous fallback
    /// </summary>
    public static CallerIdentity DefaultExtract(HttpRequest request)
    {
        // 1. Bearer token
        var auth = request.Headers.Authorization.FirstOrDefault();
        if (auth != null && auth.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
        {
            var token = auth["Bearer ".Length..];
            var tokenHash = Sha256Hex(token);
            return new CallerIdentity
            {
                Subject = $"bearer:{tokenHash[..16]}",
                AuthMethod = AuthMethod.Bearer(tokenHash),
            };
        }

        // 2. API key
        foreach (var keyHeader in new[] { "X-API-Key", "X-Api-Key", "x-api-key" })
        {
            if (request.Headers.TryGetValue(keyHeader, out var keyValues))
            {
                var keyValue = keyValues.FirstOrDefault();
                if (!string.IsNullOrEmpty(keyValue))
                {
                    var keyHash = Sha256Hex(keyValue);
                    return new CallerIdentity
                    {
                        Subject = $"apikey:{keyHash[..16]}",
                        AuthMethod = AuthMethod.ApiKey(keyHeader, keyHash),
                    };
                }
            }
        }

        // 3. Cookie
        if (request.Cookies.Count > 0)
        {
            var cookie = request.Cookies.First();
            var cookieHash = Sha256Hex(cookie.Value);
            return new CallerIdentity
            {
                Subject = $"cookie:{cookieHash[..16]}",
                AuthMethod = new AuthMethod
                {
                    Method = "cookie",
                    CookieName = cookie.Key,
                    CookieHash = cookieHash,
                },
            };
        }

        // 4. Anonymous
        return CallerIdentity.CreateAnonymous();
    }
}

File: sdks/jvm/arc-spring-boot/src/main/kotlin/io/backbay/arc/ArcIdentityExtractor.kt
/**
 * Default identity extraction from HTTP request headers.
 *
 * Mirrors the Rust extract_caller logic in arc-api-protect.
 * Extracts caller identity from Authorization headers, API keys, and cookies.
 */
package io.backbay.arc

import jakarta.servlet.http.HttpServletRequest
import java.security.MessageDigest

/** Compute SHA-256 hex digest of a string. */
fun sha256Hex(input: String): String {
    return sha256Hex(input.toByteArray(Charsets.UTF_8))
}

/** Compute SHA-256 hex digest of raw bytes. */
fun sha256Hex(input: ByteArray): String {
    val digest = MessageDigest.getInstance("SHA-256")
    val hash = digest.digest(input)
    return hash.joinToString("") { "%02x".format(it) }
}

/** Function type for extracting caller identity from a request. */
typealias IdentityExtractorFn = (HttpServletRequest) -> CallerIdentity

/**
 * Default identity extractor. Checks headers in order:
 * 1. Authorization: Bearer <token>
 * 2. X-API-Key header
 * 3. Cookie header
 * 4. Anonymous fallback
 */
fun defaultIdentityExtractor(request: HttpServletRequest): CallerIdentity {
    // 1. Bearer token
    val auth = request.getHeader("Authorization")
    if (auth != null && auth.startsWith("Bearer ", ignoreCase = true)) {
        val token = auth.removePrefix("Bearer ").removePrefix("bearer ")
        val tokenHash = sha256Hex(token)
        return CallerIdentity(
            subject = "bearer:${tokenHash.take(16)}",
            authMethod = AuthMethod.bearer(tokenHash),
        )
    }

    // 2. API key
    for (keyHeader in listOf("X-API-Key", "X-Api-Key", "x-api-key")) {
        val keyValue = request.getHeader(keyHeader)
        if (keyValue != null) {
            val keyHash = sha256Hex(keyValue)
            return CallerIdentity(
                subject = "apikey:${keyHash.take(16)}",
                authMethod = AuthMethod.apiKey(keyHeader, keyHash),
            )
        }
    }

    // 3. Cookie
    val cookies = request.cookies
    if (cookies != null && cookies.isNotEmpty()) {
        val cookie = cookies[0]
        val cookieHash = sha256Hex(cookie.value)
        return CallerIdentity(
            subject = "cookie:${cookieHash.take(16)}",
            authMethod = AuthMethod(
                method = "cookie",
                cookieName = cookie.name,
                cookieHash = cookieHash,
            ),
        )
    }

    // 4. Anonymous
    return CallerIdentity.anonymous()
}
